src/lowering/desugar.ml

```diff
@@ -386,32 +403,39 @@ and build_actor at ts self_id es obj_typ =
   let meta =
     I.{ candid = candid;
         sig_ = T.string_of_stab_sig sig_} in
-  let (interface_d, interface_f) = export_interface candid.I.service in
-  I.ActorE (interface_d @ ds', interface_f @ fs,
+  let interface_d, interface_f = export_interface candid.I.service in
+  let with_stable_vars wrap =
+    let vs = fresh_vars "v" (List.map (fun f -> f.T.typ) fields) in
+    blockE
+      ((match call_system_func_opt "preupgrade" es with
+        | Some call -> [ expD (primE (I.OtherPrim "⏰rts_perform_gc⏰") [])
```

Comment:
This also triggers a GC for the _upgrade scenario_. I guess this is what we actually want.

Reply:
Yes, I think so, though I'd prefer it if that call to perform gc was actually in canister_preupgrade, and never reachable from motoko code (that might have something on the stack - actually, aren't the values in `vs` on the stack here - I'd be nervous that `rts_perform_gc` might collect them!)

Reply:
I don't see yet how Motoko code can reach this. I am constructing a top-level actor method programmatically here.